core/src/main/java/com/google/cloud/sql/core/LazyRefreshStrategy.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import com.google.common.util.concurrent.RateLimiter;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.time.Instant;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Refresher that implements the lazy refresh strategy. */
+public class LazyRefreshStrategy implements RefreshStrategy {
+  private final Logger logger = LoggerFactory.getLogger(LazyRefreshStrategy.class);
+
+  private final String name;
+  private final Supplier<ConnectionInfo> refreshOperation;
+
+  private final RateLimiter rateLimiter;
+
+  private final Object connectionInfoGuard = new Object();
+
+  @GuardedBy("connectionInfoGuard")
+  private ConnectionInfo connectionInfo;
+
+  @GuardedBy("connectionInfoGuard")
+  private boolean closed;
+
+  /** Creates a new LazyRefresher instance. */
+  public LazyRefreshStrategy(
+      String name, Supplier<ConnectionInfo> refreshOperation, RateLimiter rateLimiter) {
+    this.name = name;
+    this.refreshOperation = refreshOperation;
+    this.rateLimiter = rateLimiter;
+  }
+
+  @Override
+  public ConnectionInfo getConnectionInfo(long timeoutMs) {
+    synchronized (connectionInfoGuard) {
+      if (closed) {
+        throw new IllegalStateException("Named connection closed");
+      }
+
+      if (connectionInfo == null) {
+        logger.debug(
+            String.format(
+                "[%s] Client certificate has expired. Starting next refresh operation immediately.",
+                name));
+        fetchConnectionInfo(timeoutMs);
+      }
+      if (Instant.now().isAfter(connectionInfo.getExpiration())) {
+        logger.debug(
+            String.format(
+                "[%s] Client certificate has expired. Starting next refresh operation immediately.",
+                name));
+        fetchConnectionInfo(timeoutMs);
+      }
+      return connectionInfo;
+    }
+  }
+
+  private void fetchConnectionInfo(long timeoutMs) {
+    synchronized (connectionInfoGuard) {
+      final long finishBeforeTimestamp = System.currentTimeMillis() + timeoutMs;
```

Comment:
Why the loop and rate limiter here? I would expect to see just a simple call and caching of the result. If we want to add retries, we should do it down below in the code that actually calls the API.

Reply:
It is easier to keep the separation of concerns consistent when the retry and rate limit are in the RefreshStrategy, not the `ConnectionInfoRepository`.

The `ConnectionInfoRepository` responsibility is simply to attempt to get the latest information from the API. If it fails, then it does not attempt to retry. 

The `RefreshStrategy` responsibility is to handle timing and concurrency. This includes the refresh delays, rate limits, and retries, which all interact with each other. By isolating all timing and concurrency control in the RefreshStrategy, we can unit test and verify it. 

The existing `RefreshAheadStrategy` implementation handles refresh delay, rate limit, and retry with Futures. This gets really weird so for legibility, it needs to all be in one place. 

The `LazyRefreshStrategy` implementation is much simpler because it is synchronous, but needs to have the same set of responsibilities: delay, rate limit, retry.

Reply:
I could see how retries being here makes sense, but the rate limiter probably doesn't belong here. If you look at the reference implementation in Go, we rate limit in the refresh ahead cache, but not in the lazy cache or the underlying API client that includes retries.

Reply:
You are correct that this forces us to implement retries (and rate limits) twice. But that is necessary because of this Java language constraint:  In Java, the waiting logic for retry and rate limit logic is implemented differently for Futures in RefreshAheadStrategy than for synchronous calls in LazyRefreshStrategy. 

The Futures implementation releases the limited thread pool threads while it waits for a retry or to acquire a rate limiter lease.  Because the Go implementation uses goroutines, the go connector can use the same implementation using blocking calls for both the Refresh Ahead and the Lazy Refresh strategies.